docs(core): add v2 plugin documentation (#18890)
nrwl/nx: docs/shared/recipes/plugins/project-graph-plugins.md

}
```

## Adding New Nodes to the Project Graph

You can add nodes to the project graph with `createNodes`. This is the API that Nx uses under the hood to identify Nx projects coming from a `project.json` file or a `package.json` that's listed in a package manager's workspaces section.

> The `createNodes` tuple is defined as follows:

Reviewer: I would link to the actual reference documentation (/packages/devkit/documents/CreateNodesFunction) since that allows them to explore all the subtypes and will always be in sync with the actual code.

Project nodes in the graph are considered to be the same if the project has the same root. If multiple plugins identify a project with the same root, the project will be merged. In doing so, the name that is already present in the graph is kept, and the properties below are shallowly merged. Any other properties are overwritten.

- `targets`
- `tags`
- `implicitDependencies`
- `generators`

Note: This is a shallow merge, so if you have a target with the same name in both plugins, the target from the second plugin will overwrite the target from the first plugin. Options, configurations, or any other properties within the target will be overwritten **_not_** merged.

#### Example

> A simplified version of the project.json plugin is shown below, which adds a new project to the project graph for each `project.json` file it finds.

Reviewer: What are they supposed to do with the `createNodes` variable?  How do they register it with Nx?

Project nodes in the graph are considered to be the same if the project has the same root. If multiple plugins identify a project with the same root, the project will be merged. In doing so, the name that is already present in the graph is kept, and the properties below are shallowly merged. Any other properties are overwritten.

- `targets`
- `tags`
- `implicitDependencies`
- `generators`

Note: This is a shallow merge, so if you have a target with the same name in both plugins, the target from the second plugin will overwrite the target from the first plugin. Options, configurations, or any other properties within the target will be overwritten **_not_** merged.

#### Example

> A simplified version of the project.json plugin is shown below, which adds a new project to the project graph for each `project.json` file it finds. This should be exported from the entry point of your plugin, which is listed in `nx.json`

Reviewer: ```suggestion
A simplified version of Nx's built-in `project.json` plugin is shown below, which adds a new project to the project graph for each `project.json` file it finds. This should be exported from the entry point file of your plugin, and the plugin should be listed under the `plugins` property of the `nx.json` file.
```

> ```typescript

Reviewer: ```suggestion
```typescript {% fileName="/my-plugin/index.ts" %}
```

  sourceFile: 'libs/existing-project/src/index.ts',
  dependencyType: DependencyType.static,
}
```

If a file hasn't changed since the last invocation, it doesn't need to be reanalyzed. Nx knows what dependencies are associated with what files, so it will reuse this information for the files that haven't changed.

### Dynamic Dependencies

Dynamic dependencies are a special type of explicit dependencies. In contrast to standard `explicit` dependencies, they are only imported in the runtime under specific conditions.
A typical example would be lazy-loaded routes. Having separation between these two allows us to identify situations where static import breaks the lazy-loading.

```typescript
{
  source: 'existing-project',
  target: 'new-project',
  sourceFile: 'libs/existing-project/src/index.ts',
  dependencyType: DependencyType.dynamic,
}
```

### Example

{% callout type="note" title="More details" %}
Even though the plugin is written in JavaScript, resolving dependencies of different languages will probably be more easily written in their native language. Therefore, a common approach is to spawn a new process and communicate via IPC or `stdout`.
{% /callout %}

> A small plugin that recognizes dependencies to projects in the current workspace which a referenced in other project's package.json files may look like so:

Reviewer: ```suggestion
A small plugin that recognizes dependencies to projects in the current workspace which are referenced in another project's `package.json` file may look like so:
```

    if (existsSync(maybePackageJsonPath)) {
      const json = JSON.parse(maybePackageJsonPath);
      packageJsonProjectMap.set(json.name, project.name);
    }
  }
  for (const project of nxProjects) {
    const maybePackageJsonPath = join(project.root, 'package.json');
    if (existsSync(maybePackageJsonPath)) {
      const json = JSON.parse(maybePackageJsonPath);
      const deps = [...Object.keys(json.dependencies)];
      for (const dep of deps) {
        if (packageJsonProjectMap.has(dep)) {
          const newDependency = {
            source: project,
            target: packageJsonProjectMap.get(dep),
            sourceFile: maybePackageJsonPath,
            dependencyType: DependencyType.static,
          };
        }
        validateDependency(ctx.graph, newDependency);
        results.push(newDependency);
      }
    }
  }
  return results;
};
```

Breaking down this example, we can see that it follows this flow:

1. Initializes an array to hold dependencies it locates
2. Builds a map of all projects in the workspace, mapping the name inside their package.json to their Nx project name.
3. Looks at the package.json files within the workspace and:

> 4. Checks if the dep is another project

Reviewer: ```suggestion
4. Checks if the dependency is another project
```
